File: scala/src/main/scala/com/ngt/streamingwithflink/util/SensorSource.scala
package com.ngt.streamingwithflink.util

import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, SourceFunction}

import java.util.Calendar
import java.util.concurrent.TimeUnit
import scala.util.Random

/**
 * @author ngt
 * @create 2021-05-07 21:29
 *         继承富函数自定义数据源
 */
class SensorSource extends RichParallelSourceFunction[SensorReading] {

  // 用于标识源代码是否继续运行
  var running: Boolean = true

  override def run(sourceContext: SourceFunction.SourceContext[SensorReading]): Unit = {
    val rand = new Random()
    val taskIdx: Int = getRuntimeContext.getIndexOfThisSubtask
    var curFTemp = (1 to 10).map {
      i => ("sensor_" + (taskIdx * 10 + i), 65 + (rand.nextGaussian() * 20))
    }

    while (running) {
      curFTemp = curFTemp.map(t => (t._1, t._2 + (rand.nextGaussian() * 0.5)))
      val curTime: Long = Calendar.getInstance.getTimeInMillis
      curFTemp.foreach(t => sourceContext.collect(SensorReading(t._1, curTime, t._2)))

      TimeUnit.MICROSECONDS.sleep(100)
    }
  }

  override def cancel(): Unit = running = false
}

File: java/src/main/java/com/ngt/streamingwithflink/util/SensorSource.java
package com.ngt.streamingwithflink.util;

import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;

import java.util.Calendar;
import java.util.Random;
import java.util.concurrent.TimeUnit;

/**
 * @author ngt
 * @create 2021-05-17 20:53
 */
public class SensorSource extends RichParallelSourceFunction<SensorReading> {
    private boolean running = true;

    @Override
    public void run(SourceContext<SensorReading> ctx) throws Exception {
        Random rand = new Random();
        int taskIdx = getRuntimeContext().getIndexOfThisSubtask();
        String[] sensorIds = new String[10];
        double[] curFTemp = new double[10];
        for (int i = 0; i < 10; i++) {
            sensorIds[i] = "sensor_" + (taskIdx * 10 + i);
            curFTemp[i] = 65 + (rand.nextGaussian() * 20);
        }

        while (running) {
            long curTime = Calendar.getInstance().getTimeInMillis();
            for (int i = 0; i < 10; i++) {
                curFTemp[i] += rand.nextGaussian() * 0.5;
                ctx.collect(new SensorReading(sensorIds[i], curTime, curFTemp[i]));
            }
            TimeUnit.MICROSECONDS.sleep(100);
        }
    }

    @Override
    public void cancel() {
        running = false;
    }
}
